workflows/mapping/Snakefile

```diff
@@ -0,0 +1,266 @@
+ALIGNER = config['rules']['align']['aligner']
+ALIGNER_PREFIX = config['rules']['align']['prefix']
+ALIGNER_TAG= {'bowtie2': 'bt2',  'hisat2': 'ht2'}[ALIGNER]
+
+if ALIGNER == 'hisat2':
+    ALIGNER_INDEX = expand('{data_dir}/{prefix}.{n}.ht2',
+                           data_dir=config['data_dir'], prefix=ALIGNER_PREFIX,
+                           n=range(1, 9))
+
+if ALIGNER == 'bowtie2':
+    ALIGNER_INDEX = expand('{data_dir}/{prefix}.{n}.bt2',
+                           data_dir=config['data_dir'], prefix=ALIGNER_PREFIX,
+                           n=range(1, 5))
+
+# ----------------------------------------------------------------------------
+# Generate final targets
+#
+patterns = [
+    '{sample_dir}/{sample}/{sample}_R1.fastqc.html',
+    '{sample_dir}/{sample}/{sample}_R1.fastq_screen.txt',
+    '{sample_dir}/{sample}/{sample}_R1.fastq_screen.png',
+    '{sample_dir}/{sample}/{sample}_R1.cutadapt.fastqc.html',
+    '{sample_dir}/{sample}/{sample}_R1.cutadapt.fastq.gz',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.bam',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.unique.bam.observed_complexity',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.unique.bam.estimated_complexity',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.unique.bam.estimated_coverage',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.unique.sort.dedup.bam',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.unique.sort.dedup.metrics',
+    'multiqc_report.html',
+    'fastq_screen_config.txt',
+    fastq_screen_config_inputs(),
+    ALIGNER_INDEX,
+]
+supplied = {
+    'sample': samples,
+    'sample_dir': config['sample_dir'],
+    'aligner_tag': ALIGNER_TAG,
+}
+
+# Fill in each pattern with each item in "supplied", resulting in a list of
+# lists. Collapse that into a list of strings, and then unique them into a set
+# since e.g. expand('a', samples=[1, 2, 3, 4]) results in ['a', 'a', 'a', 'a']
+targets = list(
+    set(
+        itertools.chain(
+            *map(lambda x: expand(x, **supplied), patterns)
+        )
+    )
+)
+# ----------------------------------------------------------------------------
+
+
+
+rule all:
+    input: targets
+
+
+rule clean:
+    run:
+        shell('rm -r {config[sample_dir]}')
+        shell('./get-data.sh')
+
+
+rule fastqc:
+    input: 'pasilla/{sample}/{sample}{suffix}.fastq.gz'
+    output:
+        html='pasilla/{sample}/{sample}{suffix}.fastqc.html',
+        zip='pasilla/{sample}/{sample}{suffix}_fastqc.zip',
+    params: extra=params_for('fastqc', 'extra')
+    wrapper:
+        wrapper_for('fastqc')
+
+
+rule cutadapt:
+    input: 'pasilla/{sample}/{sample}_R1.fastq.gz'
+    output: 'pasilla/{sample}/{sample}_R1.cutadapt.fastq.gz'
+    log: 'pasilla/{sample}/{sample}_R1.cutadapt.fastq.log'
+    params: extra=params_for('cutadapt', 'extra')
+    wrapper:
+        wrapper_for('cutadapt')
+
```

Comment:
Kind of going back to snakemake-lib, one nice features was the separation files into distinct levels:

* raw level
* run level
* sample level

Maybe we could introduce some of this hierarchy so instead of `prefix` may have something like `run_prefix`, `sample_prefix`, `agg_prefix`, or even `chip_prefix`. Then maybe we could come up with some cute input function that can jump up or down levels. 

Just a thought, this would allow you to quickly change folder structure with minor edits or even a config option.